text_metrics/utils.py

```diff
@@ -271,7 +282,242 @@ def _join_surp(words: list[str], tok_surps: list[tuple[str, float]]):
+        except:
+            max_ctx = 8192
+        # print(max_ctx)
+        while True:
+            encodings = tokenizer(
+                sentence[start_ind:],
+                max_length=max_ctx - 2,
+                truncation=True,
+                return_offsets_mapping=True,
+            )
+            # for gpt and pythia variants, we need to add bos and eos tokens
+            if (
+                "gpt-neox" in model_variant
+                or "opt" in model_variant
+                or "mamba" in model_variant
+                or "Llama" in model_variant
+            ):
+                tensor_input = torch.tensor(
+                    [encodings["input_ids"] + [tokenizer.eos_token_id]],
+                    device=model.device,
+                )
+            elif "gpt" in model_variant or "pythia" in model_variant:
+                tensor_input = torch.tensor(
+                    [
+                        [tokenizer.bos_token_id]
+                        + encodings["input_ids"]
+                        + [tokenizer.eos_token_id]
+                    ],
+                    device=model.device,
+                )
+            else:
+                raise ValueError(f"Unsupported LLM variant {model_variant}")
+
+            output = model(tensor_input, labels=tensor_input)
+            shift_logits = output["logits"][
+                ..., :-1, :
+            ].contiguous()  # remove the last token from the logits
+
+            #  This shift is necessary because the labels are shifted by one position to the right
+            # (because the logits are w.r.t the next token)
+            shift_labels = tensor_input[
+                ..., 1:
+            ].contiguous()  # remove the first token from the labels,
+
+            log_probs = torch.nn.functional.cross_entropy(
+                shift_logits.view(-1, shift_logits.size(-1)),
+                shift_labels.view(-1),
+                reduction="none",
+            )
+
+            # varify that the average of the log_probs is equal to the loss
+            assert torch.isclose(
+                torch.exp(sum(log_probs) / len(log_probs)), torch.exp(output["loss"])
+            )
+
+            # Handle the case where the sentence is too long for the model
+            offset = 0 if start_ind == 0 else stride - 1
+            all_log_probs = torch.cat([all_log_probs, log_probs[offset:-1]])
+            offset_mapping.extend(
+                [
+                    (i + start_ind, j + start_ind)
+                    for i, j in encodings["offset_mapping"][offset:]
+                ]
+            )
+            if encodings["offset_mapping"][-1][1] + start_ind == len(sentence):
+                break
+            start_ind += encodings["offset_mapping"][-stride][1]
+
+    if (
+        "gpt-neox" in model_variant
+        or "opt" in model_variant
+        or "mamba" in model_variant
+        or "Llama" in model_variant
+    ):
```

Comment:
```suggestion
    if any(variant in model_variant for variant in ["gpt-neox", "opt", "mamba", "Llama"]):
```

Reply:
Maybe define a list and reuse in both places? Also consider lower case, the "Ll" in "Llama" is scary